spec.md

```diff
@@ -2734,6 +2790,257 @@ message NodeExpandVolumeResponse {
+  // If any snapshot in the list of snapshots in this message have
+  // ready_to_use set to false, the SP MUST set this field to false.
+  // If all of the snapshots in the list of snapshots in this message
+  // have ready_to_use set to true, the SP SHOULD set this field to
+  // true.
+  // This field is REQUIRED.
+  bool ready_to_use = 4;
+}
+```
+
+##### CreateVolumeGroupSnapshot Errors
+
+If the plugin is unable to complete the CreateVolumeGroupSnapshot call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Group snapshot already exists but is incompatible | 6 ALREADY_EXISTS | Indicates that a group snapshot corresponding to the specified group snapshot `name` already exists but is incompatible with the specified `volume_id`. | Caller MUST fix the arguments or use a different `name` before retrying. |
+| Cannot snapshot multiple volumes together | 9 FAILED_PRECONDITION | Indicates that the specified volumes cannot be snapshotted together because the volumes are not configured properly based on requirements from the SP. | Caller MUST fix the configuration of the volumes so that they meet the requirements for group snapshotting before retrying. |
+| Not enough space to create group snapshot | 13 RESOURCE_EXHAUSTED | There is not enough space on the storage system to handle the create group snapshot request. | Caller SHOULD fail this request. Future calls to CreateVolumeGroupSnapshot MAY succeed if space is freed up. |
+
+#### `DeleteVolumeGroupSnapshot`
+
+**ALPHA FEATURE**
+
+A Controller Plugin MUST implement this RPC call if it has `CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT` capability.
+This RPC will be called by the CO to delete a volume group snapshot.
+This operation will delete a volume group snapshot as well as all individual snapshots that are part of this volume group snapshot.
+
+This operation MUST be idempotent.
+If a group snapshot corresponding to the specified `group_snapshot_id` does not exist or the artifacts associated with the group snapshot do not exist anymore, the Plugin MUST reply `0 OK`.
+
+```protobuf
+message DeleteVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // The ID of the group snapshot to be deleted.
+  // This field is REQUIRED.
+  string group_snapshot_id = 1;
+
+  // A list of snapshot IDs that are part of this group snapshot.
+  // If SP does not need to rely on this field to delete the snapshots
+  // in the group, this field MAY be ignored.
+  // Some SPs require this list to delete the snapshots in the group.
+  // If SP needs to use this field to delete the snapshots in the
+  // group, it MAY report an error when there is a mismatch.
```

Comment:
if the SP needs this field to properly execute the operation, and there is a mismatch between the IDs here and the group ID then .. "MAY" seems like a strange choice of constraint. I imagine that the results of such a mismatch in this case fall into one of the following buckets: (A) there is an undetected resource leak (because an individual ID was missing); (B) there is an accidental delete of an individual snapshot (because some individual ID was not actually part of the group); (C) the operation fails because some internal constraint isn't met when individual IDs don't align w/ the group ID (e.g. some internal checksum mismatch when IDs don't completely correlate; the "group bucket" for snapshot data can't be found if exact IDs don't match, etc).

I think this needs to be a lot tighter, in terms of SPs generating error codes when there's a mismatch between individual IDs and group ID. the SP *MUST* be able to detect a mismatch between an individual ID and group ID (if it requires individual IDs in this message). note that the requirements for detecting a mismatch are easier to meet vs. requiring the SP to store a 1:1 mapping between group ID and individual IDs.

if the SP truly doesn't care about this field, because it doesn't need the individual IDs, then it MAY ignore the field - but probably SHOULD log a mismatch if it's able to detect such. because the way that the spec is written, a CO is *very* likely to provide this field every time for this call in order to maximize compat across all SPs (and it's in everyone's interest to detect alignment mismatches).

also, if it returns an error due to a mismatch then how is that reported back to the CO? and what's the mitigation? (there is nothing in the errors table describing this scenario)

Reply:
Updated the comments to the following:
```
  // If SP does not need to rely on this field to delete the snapshots
  // in the group, it SHOULD check this field and report an error
  // if it has the ability to detect a mismatch.
  // Some SPs require this list to delete the snapshots in the group.
  // If SP needs to use this field to delete the snapshots in the
  // group, it MUST report an error if it has the ability to detect
  // a mismatch.
```
Added a INVALID_ARGUMENT error code for the snapshot id mismatch error.

Reply:
Addressed the comment here: https://github.com/container-storage-interface/spec/pull/519/commits/8959d25967193aa050f15b0b76f6f6bf0d69d5b6